implement a very basic TLS 1.3 handshake

marten-seemann opened the pull request.

lucas-clemente commented on crypto/aesgcm_aead.go:
```diff
@@ -17,31 +16,34 @@ type aeadAESGCM struct {
+		return nil, errors.New("AES-GCM: expected 12 byte IVs")
 	}
-	encrypterCipher, err := aes12.NewCipher(myKey)
+
```
Is there a reason why you're not checking the key size here (anymore)?

marten-seemann replied:
Yes, `aes.NewCipher` can be called with different key sizes, so it performs the check if the key has one of the accepted sizes. See https://golang.org/pkg/crypto/aes/#NewCipher

lucas-clemente replied:
Okay, and mint will check for us that we don't accept weak keys?

marten-seemann replied:
TLS 1.3 doesn't have any weak cipher suites. These are the ones supported by mint: https://github.com/bifurcation/mint/blob/master/common.go#L43-L52.

marten-seemann replied:
I added tests with 16 and 32 byte keys, and another test that we error on invalid key sizes.

lucas-clemente commented on crypto/key_derivation.go:
```diff
@@ -0,0 +1,49 @@
+	serverExporterLabel = "EXPORTER-QUIC server 1-RTT Secret"
+)
+
+// MintState contains
```
Nit

lucas-clemente commented on crypto/key_derivation.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+func computeKeyAndIV(ms MintState, label string) (key, iv []byte, err error) {
+	cs := ms.GetCipherSuite()
```
Should we check here that this is actually AES-GCM?

marten-seemann replied:
We probably should do that. That will mean checking the cipher suite against the values defined in https://github.com/bifurcation/mint/blob/master/common.go#L43-L52 though, I don't see a general way to check for AES.

marten-seemann replied:
Actually, what we should do is set a list of supported ciphers in the `mint.Config`, and only use AES ciphers there. We can then trust mint to negotiate one of them.

lucas-clemente commented on handshake/crypto_setup_tls.go:
```diff
@@ -0,0 +1,140 @@
+		perspective:   perspective,
+		mintConf:      mintConf,
+		conn:          conn,
+		nullAEAD:      crypto.NewNullAEAD(perspective, protocol.Version39),
```
Why is there a version const here?

marten-seemann replied:
I think I forgot to write the correct null AEAD constructor. I will add this later today.

marten-seemann replied:
Fixed.

lucas-clemente commented on handshake/crypto_setup_tls.go:
```diff
@@ -0,0 +1,140 @@
+
+func (h *cryptoSetupServerTLS) HandleCryptoStream() error {
+	alert := h.conn.Handshake()
+	utils.Debugf("ALERT: %#v", alert)
```
Remove this line or move it into the if block below.

marten-seemann commented:
@lucas-clemente: I think I addressed all your comments. Furthermore, I moved all mint-related methods into one interface, which should make testing easier. PHAL.
